src/Backend/SgateShopgatePlugin/Controllers/Frontend/shopgate.php

```diff
@@ -288,4 +377,988 @@ public function pluginAction()
+        $segments = explode('/', trim($path, '/'));
+
+        $sessionId = $segments[2];
+
+        if (isset($sessionId)) {
+            session_write_close();
+            session_id($sessionId);
+            session_start([
+                'sessionId' => $sessionId
+            ]);
+        }
+
+        $sgCloud = $this->Request()->getParam('sgcloud_checkout');
+
+        $this->session->offsetSet('sgWebView', true);
+        $this->Response()->setHeader('Set-Cookie', 'sgWebView=true; path=/; HttpOnly');
+
+        if(isset($sgCloud)) {
+            $this->Response()->setHeader('Set-Cookie', 'session-1='.$sessionId.'; path=/; HttpOnly');
+            $this->redirect('checkout/confirm#show-registration');
+        } else {
+            $this->redirect('account#show-registration');
+        }
+    }
+
+    protected function verifyUser($email, $hash)
+    {
+        if (empty($email)) {
+            $sErrorFlag['email'] = true;
+        }
+        if (empty($hash)) {
+            $sErrorFlag['password'] = true;
+        }
+
+        $addScopeSql = '';
+        if ($this->scopedRegistration == true) {
+            $addScopeSql = $this->db->quoteInto(' AND subshopID = ? ', $this->subshopId);
+        }
+
+        $preHashedSql = $this->db->quoteInto(' AND password = ? ', $hash);
+
+        $sql = '
+                SELECT id, customergroup, password, encoder
+                FROM s_user WHERE email = ? AND active=1
+                AND (lockeduntil < now() OR lockeduntil IS NULL) '
+            . $addScopeSql
+            . $preHashedSql;
+
+        $getUser = $this->db->fetchRow($sql, [$email]) ?: [];
+
+        if (!count($getUser)) {
+            $isValidLogin = false;
+        } else {
+            $encoderName = 'Prehashed';
+
+            $plaintext = $hash;
+            $password = $getUser['password'];
+
+            $isValidLogin = $this->passwordEncoder->isPasswordValid($plaintext, $password, $encoderName);
+        }
+
+        if (!$isValidLogin) {
+            $sErrorMessages = [];
+            $sErrorMessages['sErrorMessages'] = 'your account is invalid';
+            return $sErrorMessages;
+        }
+
+        $userId = $getUser['id'];
+        $sql = '
+            SELECT * FROM s_user
+            WHERE password = ? AND email = ? AND id = ?
+            AND UNIX_TIMESTAMP(lastlogin) >= (UNIX_TIMESTAMP(now())-?)
+        ';
+
+        $user = $this->db->fetchRow(
+            $sql, [$hash, $email, $userId, 7200,]
+        );
+
+        return $user;
+    }
+
+    /**
+     * @return Shopware_Plugins_Backend_SgateShopgatePlugin_Components_Config
+     */
+    protected function getConfig()
+    {
+        static $config = null;
+
+        if (!$config) {
+            $config = new Shopware_Plugins_Backend_SgateShopgatePlugin_Components_Config();
+        }
+
+        return $config;
+    }
+
+    /**
+     * WARNING: ALL functions listed below have been directly copied from Shopware's Checkout or Basket controller.
```

Comment:
not quite true as the JWT added at the bottom has no relationship to the controller.